multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,67 +78,215 @@ func NewNodeReconciler(
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      req.Name,
 			Namespace: r.namespace,
 		},
 	}
-	// When the Node is annotated with 'multicluster.antrea.io/gateway=true' as a Gateway:
-	//   - Delete the Gateway if the Node is deleted
-	//   - Update the Gateway if Node's InternalIP or GatewayIP is updated
-	//   - Create a new Gateway if there is no existing Gateway
+
 	node := &corev1.Node{}
 	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
 		if apierrors.IsNotFound(err) {
-			err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
-			return ctrl.Result{}, client.IgnoreNotFound(err)
+			if r.activeGateway == req.Name || r.activeGateway == "" {
+				if err := r.recreateActiveGateway(ctx, gw); err != nil {
+					return ctrl.Result{}, err
+				}
+				return ctrl.Result{}, nil
+			}
 		}
 		klog.ErrorS(err, "Failed to get Node", "node", req.Name)
-		return ctrl.Result{}, err
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	_, isGW := node.Annotations[common.GatewayAnnotation]
-	var err error
-	gwNamespacedName := types.NamespacedName{
-		Name:      node.Name,
-		Namespace: r.namespace,
-	}
+	_, hasGWAnnotation := node.Annotations[common.GatewayAnnotation]
+	nodeIsReady := isReadyNode(node)
+	noActiveGateway := r.activeGateway == ""
+	isActiveGateway := r.activeGateway == node.Name
 
-	// TODO: cache might be stale. Need to revisit here and other reconcilers to
-	// check if we can improve this with 'Owns' or other methods.
-	var gwIP, internalIP string
-	if isGW {
-		if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
-			klog.ErrorS(err, "There is no valid Gateway IP for Node, will retry later when there is any new Node update", "node", node.Name)
+	if !nodeIsReady {
+		if isActiveGateway {
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
+				return ctrl.Result{}, err
+			}
 			return ctrl.Result{}, nil
 		}
+		if hasGWAnnotation {
+			r.gatewayCandidates[req.Name] = req.Name
+		}
+		return ctrl.Result{}, nil
 	}
-	if err := r.Client.Get(ctx, gwNamespacedName, gw); err != nil {
-		if apierrors.IsNotFound(err) && isGW {
-			gw.GatewayIP = gwIP
-			gw.InternalIP = internalIP
-			if err := r.Client.Create(ctx, gw, &client.CreateOptions{}); err != nil {
+
+	var err error
+	var gwIP, internalIP string
+	// When Node is ready without Gateway annotation
+	if !hasGWAnnotation {
+		if isActiveGateway {
+			// The annotation is removed from the Node, we need recreate a new Gateway.
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
 				return ctrl.Result{}, err
 			}
-			return ctrl.Result{}, nil
 		}
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		return ctrl.Result{}, nil
 	}
 
-	if !isGW {
-		err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+	// When the Node is ready with the Gateway annotation
+	r.gatewayCandidates[node.Name] = node.Name
+	internalIP, gwIP, err = r.getGatawayNodeIP(node)
```

Comment:
If would check all conditions of recreateActiveGateway() and call it once, rather than in three places for three conditions.

Reply:
Good point, refined, thanks.